File: pymultisig/btc_utils.py

from bitcoin.core import script


def estimate_transaction_fees(num_inputs: int,
                              num_outputs: int,
                              fees_per_byte_in_satoshis: int) -> int:
    """
    Estimate total transaction fee for non-segwit, 2-of-3 P2SH-multisig UTXO spends

    Fee estimation is based of the following encoding assumptions
    10 Overhead bytes:
      version - 4
      input count - 1
      output count - 1
      locktime - 4
    
    297 Bytes per 2/3-multisig utxo:
      prev_tx - 32
      index -4
      scriptsize - 3
      scriptsig - 254
      sequence - 4
    
    34 Bytes per output (conservative assuming P2PKH)

    Args:
        num_inputs: number of UTXOs being spent
        num_outputs: number of outputs
        fees_per_byte_in_satoshis:

    Returns:
        Total transaction fee in satoshis

    Example:
        TODO
    """
    
    tx_length_in_bytes = 10 + (297 * num_inputs) + (34 * num_outputs)

    fees_in_satoshis   = tx_length_in_bytes * fees_per_byte_in_satoshis

    return fees_in_satoshis


def parse_redeem_script(redeemscript: str) -> dict:
    """
    Parses a 2-of-3 P2SH-multisig redeem script

    Args:
        redeemscript: hex-encoded redeem script

    Returns:
        Dictionary with two keys:
            m: M for M-of-N multisig
            pubkeys: three hex-encoded compress public keys

    Raises:
        ValueError: If the redeem script isn't valid 2-of-3 script

    Example:
        TODO
    """
    
    parsed = script.CScript.fromhex(redeemscript)

    parsed_ops = tuple([i if (type(i) == script.CScriptOp or type(i) == int) else type(i) for i in parsed])

    parsed_items = [x for x in parsed]

    multisig_template = (2, bytes, bytes, bytes, 3, script.OP_CHECKMULTISIG)
    
    if parsed_ops != multisig_template:
        raise ValueError("Redeem Script is not Valid 2-of-3 Multisig")
    else:
        return {"m": parsed_items[0],
                "pubkeys": [parsed_items[1].hex(),
                            parsed_items[2].hex(),
                            parsed_items[3].hex()
                            ]
                }
    

File: src/utils/bitcoin/btc_utils.js
const bitcoin = require('bitcoinjs-lib');


function validateHex(inputString) {
    if (inputString.length % 2) {
        throw new Error("invalid hex - odd-length string");
    }
    var re = /^[0-9A-Fa-f]*$/;
    if (!re.test(inputString)) {
        throw new Error("invalid hex - invalid characters");
    }
}



function compressPublicKey(publicKey) {
    validateHex(publicKey);
    //validate Public Key Length
    //validate Public Key Structure
    let pubkey_buffer = Buffer.from(publicKey, 'hex');
    let prefix = (pubkey_buffer[64] & 1) !== 0 ? 0x03 : 0x02;
    let prefixBuffer = Buffer.alloc(1);
    prefixBuffer[0] = prefix;
    return Buffer.concat([prefixBuffer, pubkey_buffer.slice(1, 1 + 32)]).toString('hex');
}

function generateMultisigRedeemScript(pubkey1, pubkey2, pubkey3) {
    let pubkeys = [pubkey1, pubkey2, pubkey3
                  ].map((hex) => Buffer.from(hex, 'hex'));

    let redeem = bitcoin.payments.p2ms({ m: 2, pubkeys });

    return redeem.output.toString('hex');
}

function generateMultisigAddress(redeemScript, testnet) {
    let net;
    if (testnet) {
        net = bitcoin.networks.testnet;
    } else {
        net = bitcoin.networks.bitcoin;
    }

    let redeemP2MS = bitcoin.payments.p2ms({output: Buffer.from(redeemScript, 'hex'),
                                            network: net});

    let { address } = bitcoin.payments.p2sh({
        redeem: redeemP2MS,
        network: net
    });

    return address;
}

function extractPublicKeysHex(redeemScript) {
    let parsedRedeemScript = bitcoin.payments.p2ms(
        {output: Buffer.from(redeemScript,'hex')}
    );        
    let pubKeysHex = parsedRedeemScript.pubkeys.map(x => x.toString('hex'));
    // validate M and N?
    return pubKeysHex;
}
    


function estimateTransactionFees(numInputs, numOutputs, feesPerByteInSatoshis) {
    /*
      Estimate total transaction fee for non-segwit, 2-of-3 P2SH-multisig UTXO spends

      Fee estimation is based of the following encoding assumptions
      10 Overhead bytes:
      version - 4
      input count - 1
      output count - 1
      locktime - 4
      
      297 Bytes per 2/3-multisig utxo:
      prev_tx - 32
      index -4
      scriptsize - 3
      scriptsig - 254
      sequence - 4
      
      34 Bytes per output (conservative assuming P2PKH)

      Args:
      num_inputs: number of UTXOs being spent
      num_outputs: number of outputs
      fees_per_byte_in_satoshis:

      Returns:
      Total transaction fee in satoshis

      Example:
      TODO
    */
    
    let txLengthInBytes = 10 + (297 * numInputs) + (34 * numOutputs);

    let feesInSatoshis   = txLengthInBytes * feesPerByteInSatoshis;

    return feesInSatoshis;
}




let myexports = { compressPublicKey,
                   estimateTransactionFees,
                   generateMultisigRedeemScript,
                   generateMultisigAddress,
                   extractPublicKeysHex };

(function(exports){
})(typeof exports === 'undefined'? this['btcUtils']=myexports: myexports);

if (typeof module.exports === 'object') module.exports = myexports;
